src/main/java/me/ryanhamshire/GPFlags/util/ChatUtil.java

```diff
@@ -0,0 +1,258 @@
+            boolean requiredChanges = false;
+            String clickComponent = message.substring(matcher.start(), matcher.end());
+            if (clickComponent.contains("suggest_command") && !sender.hasPermission("gpflags.messages.click.suggest_command")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            if (clickComponent.contains("run_command") && !sender.hasPermission("gpflags.messages.click.run_command")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            if (clickComponent.contains("open_url") && !sender.hasPermission("gpflags.messages.click.open_url")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            matcher = pattern.matcher(message);
+
+            if (!requiredChanges) {
+                break;
+            }
+        }
+
+        // Check for hover-components
+        pattern = Pattern.compile("<hover:.*?>");
+        matcher = pattern.matcher(message);
+        while (matcher.find()) {
+            boolean requiredChanges = false;
+
+            String hoverComponent = message.substring(matcher.start(), matcher.end());
+            if (hoverComponent.contains("show_text") && !sender.hasPermission("gpflags.messages.hover.show_text")) {
+                message = message.replace(hoverComponent, "");
+                requiredChanges = true;
+            }
+            if (hoverComponent.contains("show_item") && !sender.hasPermission("gpflags.messages.hover.show_item")) {
+                message = message.replace(hoverComponent, "");
+                requiredChanges = true;
+            }
+            if (hoverComponent.contains("show_entity") && !sender.hasPermission("gpflags.messages.hover.show_entity")) {
+                message = message.replace(hoverComponent, "");
+                requiredChanges = true;
+            }
+            matcher = pattern.matcher(message);
+
+            if (!requiredChanges) {
+                break;
+            }
+        }
+
+        // Check for hex-colors
+        message = message.replace("<#", "<color:#");
+        pattern = Pattern.compile("<color:#.*?>");
+        matcher = pattern.matcher(message);
+        while (matcher.find()) {
+            boolean requiredChanges = false;
+            String hexComponent = message.substring(matcher.start(), matcher.end());
+            if (!sender.hasPermission("gpflags.messages.hex-colors")) {
+                message = message.replace(hexComponent, "<color:white>");
+                requiredChanges = true;
+            }
+            matcher = pattern.matcher(message);
+
+            if (!requiredChanges) {
+                break;
+            }
+        }
+
+        // Check for colors
+        message = ChatUtil.parseNativeColorCodes(message);
+        pattern = Pattern.compile("<color:.*?>");
```

Comment:
It's also possible to make colors as like \<blue> or \<red> and I don't think you consider this.

Reply:
You are right, ill implement them as well

Reply:
Got it added with an even greater approach. I will directly replace all `&<color>` & `<[color]>` into the proper `<color:[color]`, but in the whole message. Meaning we save time on iterating over the chars which was done before.

Reply:
That works